### ISSUE TITLE
DNS providers fail to present token (read udp i/o timeout)

### ISSUE THREAD

**Issue by janeczku** [xenolf/lego#179]
Since #142  i am getting intermittent errors with DNS providers:

`level=error msg="[xxx.xxx.xxx] Error obtaining certificate: Error presenting token read udp xxx.xxx.xxx:57113->8.8.8.8:53: i/o timeout"`

As discussed in https://github.com/xenolf/lego/pull/142#issuecomment-189884283 the `acme.FindZoneByFqdn` function doesn't handle connection errors when sending a query to the nameserver. Since DNS uses UDP a single lost/dropped packet will cause the connection to fail. This doesn't matter when the function is already wrapped in a retry loop but #142 uses this function directly to lookup the zone used in API queries.

@LukeHandle

**Comment by LukeHandle** [comment]
As before, I suggest we add such functionality in the `dnsQuery` function.

How many times would you consider best for retrying here, would you suggest switching to TCP in the event of an issue for some more reliability?

Finally, do you suggest we only retry when the error is specifically connection issues, or is it worth retrying on any error?

**Comment by LukeHandle** [comment]
Another thing, where abouts was this error being generated? I was struggling to find the strings in the message eg. https://github.com/xenolf/lego/search?utf8=%E2%9C%93&q=Error+obtaining+certificate&type=Code

```
grep -R "certificate:" .                                                        
/acme/client.go:                                logf("[ERROR][%s] acme: Could not bundle issuer certificate: %v", cert.Domain, err)
/acme/client.go:                                                logf("[WARNING][%s] acme: Could not bundle issuer certificate: %v", commonName.Domain, err)
README.md:Obtain a certificate:
README.md:To renew the certificate:
```

**Comment by janeczku** [comment]
> How many times would you consider best for retrying here, would you suggest switching to TCP in the event of an issue for some more reliability?

1. I would rather fallback to the alternate Google DNS (8.8.4.4) when the initial query failed. DNS over TCP is not really purposed for redundancy. It's also worth noting that Google's Public DNS servers throttles clients exceeding a certain amount of queries per minute. They do this by silently dropping the request which looks like a lost packet from our side. If we are rate limited switching to TCP won't help but switching to the alternate server will.
I don't have a strong opinion on how many times we should retry. Maybe max three tries total?

> Finally, do you suggest we only retry when the error is specifically connection issues, or is it worth retrying on any error?

The only error requiring special attention is `ErrTruncated` which is already handled in the current code. Any other error will essentially be either connection related or due to corrupt DNS message data. Both are worth retrying.

> Another thing, where abouts was this error being generated?

https://github.com/xenolf/lego/blob/master/acme/dns_challenge.go#L57-L59

**Comment by janeczku** [comment]
@LukeHandle Are you working on a fix?

**Comment by LukeHandle** [comment]
@janeczku I plan to, but I cant promise anything before the weekend as I'm currently away from home. If there is urgency on this, please go ahead.

**Comment by janeczku** [comment]
@LukeHandle Thanks for the heads up!

**Comment by LukeHandle** [comment]
Hey @janeczku , are you able to take a look at #186 ?

I thought that using the alternate for both the 2nd and 3rd attempt would be simpler than attempting to switch back to the original for the third. The end result is realistically the same anyway.

Any thoughts/critiques?

**Comment by xenolf** [comment]
Closed via #186
